lib/compress/zstd_compress.c

```diff
@@ -2516,24 +2518,22 @@ static void ZSTD_copyBlockSequences(ZSTD_CCtx* zc)
         }
 
         if (seqStoreSeqs[i].offset <= ZSTD_REP_NUM) {
+            /* Derive the correct offset corresponding to a repcode */
             outSeqs[i].rep = seqStoreSeqs[i].offset;
-            repIdx = (unsigned int)i - seqStoreSeqs[i].offset;
-
-            if (seqStoreSeqs[i].litLength == 0) {
-                if (seqStoreSeqs[i].offset < 3) {
-                    --repIdx;
+            if (outSeqs[i].litLength != 0) {
+                rawOffset = rep[outSeqs[i].rep - 1];
+            } else {
+                if (outSeqs[i].rep == 3) {
+                    rawOffset = rep[0] - 1;
                 } else {
-                    repIdx = (unsigned int)i - 1;
+                    rawOffset = rep[outSeqs[i].rep];
                 }
             }
-            assert(repIdx >= -3);
-            outSeqs[i].offset = repIdx >= 0 ? outSeqs[repIdx].offset : repStartValue[-repIdx - 1];
-            if (outSeqs[i].rep == 3 && outSeqs[i].litLength == 0) {
-                --outSeqs[i].offset;
-            }
-        } else {
-            outSeqs[i].offset = seqStoreSeqs[i].offset - ZSTD_REP_NUM;
         }
+        outSeqs[i].offset = rawOffset;
+        updatedRepcodes = ZSTD_updateRep(rep, seqStoreSeqs[i].offset - 1, seqStoreSeqs[i].litLength == 0);
+        ZSTD_memcpy(rep, updatedRepcodes.rep, sizeof(repcodes_t));
```

Comment:
It would be slightly more efficient to do:
1. Copy `rep` into `updatedRepcodes` before the loop
2. Only used `updatedRepcodes` in the loop
3. Copy `updatedRepcodes` into `rep` after the loop